source/source.go

```diff
@@ -0,0 +1,267 @@
+// Copyright © 2022 Meroxa, Inc. and Miquido
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/miquido/conduit-connector-salesforce/internal/cometd"
+	"github.com/miquido/conduit-connector-salesforce/internal/cometd/responses"
+	"github.com/miquido/conduit-connector-salesforce/internal/salesforce/oauth"
+)
+
+const sfCometDVersion = "54.0"
+
+type Source struct {
+	sdk.UnimplementedSource
+
+	config          Config
+	streamingClient *cometd.Client
+	subscriptions   map[string]bool
+	events          chan responses.ConnectResponseEvent
+	errors          chan error
+}
+
+func NewSource() sdk.Source {
+	return &Source{
+		subscriptions: make(map[string]bool),
+		events:        make(chan responses.ConnectResponseEvent),
+		errors:        make(chan error),
+	}
+}
+
+func (s *Source) Configure(_ context.Context, cfgRaw map[string]string) (err error) {
+	s.config, err = ParseConfig(cfgRaw)
+
+	return
+}
+
+func (s *Source) Open(ctx context.Context, _ sdk.Position) error {
+	// Authenticate
+	oAuthClient := oauth.NewClient(
+		s.config.Environment,
+		s.config.ClientID,
+		s.config.ClientSecret,
+		s.config.Username,
+		s.config.Password,
+		s.config.SecurityToken,
+	)
+
+	token, err := oAuthClient.Authenticate(ctx)
+	if err != nil {
+		return fmt.Errorf("could not authenticate: %w", err)
+	}
+
+	// Streaming API client
+	s.streamingClient, err = cometd.NewClient(
+		fmt.Sprintf("%s/cometd/%s", token.InstanceURL, sfCometDVersion),
+		token.AccessToken,
+	)
+	if err != nil {
+		return fmt.Errorf("could not create Streaming API client: %w", err)
+	}
+
+	// Handshake
+	if _, err := s.streamingClient.Handshake(ctx); err != nil {
+		return fmt.Errorf("handshake error: %w", err)
+	}
+
+	// Subscribe to topic
+	for _, pushTopicName := range s.config.PushTopicsNames {
```

Comment:
What are the differences between the approach with push topic and this one with CDC: https://developer.salesforce.com/docs/atlas.en-us.change_data_capture.meta/change_data_capture/cdc_intro.htm ?

Reply:
I honestly don't know. I focused on a Streaming API, suggested on an initial meeting. I can check if current solution can subscribe to data topics.

Reply:
OK, I tested this. It requires slightly different configuration, but (from the source code perspective) subscribing process is the same as for Push Topics. I can implement support for this. The event payload is different.

Required change: instead of the current Push Topics Names config (e.g. `TaskUpdates,ContactUpdates`), I would need to rename it to `Channels Names` and require the user to provide the event source: `/topic/TaskUpdates,/topic/ContactUpdates,/data/ChangeEvents,/data/TaskChangeEvent,...`.

Reply:
@donatorsky Thanks for checking that! What are the differences in features of Streaming API vs CDC? Would we get more events, easier setup, historical changes etc.?

Reply:
@hariso Main changes I noticed:
- Easier and minimal setup. Just go to _Setup -> Integrations -> Change Data Capture_ and select objects You are interested in.
- You receive only changed values. With Push Topics, each change generates a full payload (depending on SELECTed fields in the topic query). However, You will receive notifications about changes You are not interested in. I guess it all depends on use case.
- You also receive non-flat data structure. Example event's payload:
```application/json
"payload": {
    "LastModifiedDate": "2022-05-18T13:50:57.000Z",
    "Name": {
        "FirstName": "Foo 2",
        "LastName": "Bar 3"
    },
    "CustomGeo__c": {
        "Latitude": 11.7,
        "Longitude": 33.2
    },
    "ChangeEventHeader": {
        "commitNumber": 11254192394289,
        "commitUser": "0058c00000BdfZEAAZ",
        "sequenceNumber": 1,
        "entityName": "Contact",
        "changeType": "UPDATE",
        "changedFields": ["LastModifiedDate", "Name.FirstName", "Name.LastName", "CustomGeo__c.Latitude", "CustomGeo__c.Longitude"],
        "changeOrigin": "com/salesforce/api/soap/54.0;client=SfdcInternalAPI/",
        "transactionKey": "0003a428-f024-7140-33b5-fa61a4150783",
        "commitTimestamp": 1652881857000,
        "recordIds": ["0038c00002lNgEfAAK"]
    }
}
```
This also involves more complicated parsing, but still can be done. More data goes to metadata section. With Push Topic, You would get something like this:
```application/json
{
    "FirstName": "Foo 2",
    "LastName": "Bar 3"
    "CustomGeo__c_Latitude": 11.7,
    "CustomGeo__c_Longitude": 33.2
}
```

Reply:
All of that is some pretty useful information. Let's please have it in the ReadMe (feel free to do it in a separate PR if you'd like to).

Reply:
Sure. For this PR scope, I'll just add information that CDC events are not supported. It's definitely a better idea to finish this initial commit and do this in a separate, smaller commit. Let me know when you decide you want to support CDC events 🙂

Reply:
BTW, I found nice Push Topics and CDC comparison:
https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/event_comparison.htm

Reply:
@donatorsky If you don't mind please create an issue for the CDC events, so we have it on the radar (and so that whoever might want to use this connector is aware that we have considered it).

> BTW, I found nice Push Topics and CDC comparison:
> https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/event_comparison.htm

Nice!

Reply:
OK. I'll include in the issue information gathered here to have it all in one place. I'm on it!